packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,199 @@
+
+Once _any_ row has been expanded or collapsed, we must update the Nimble `Table` state where we track collapsed rows with _all_ rows that are currently collapsed. This will be a one-time cost. Prototyping suggests that the performance penalty isn't that noticeable.
+
+### Showing a progress indicator for lazy loading
+
+After a user clicks on a parent row to expand it, and that parent does not have children yet, there could be a noticeable delay between when the user clicked the expand button, and when the children show up in the table. This could happen for a variety of reasons, but commonly might occur due to a slow network request.
+
+It would be helpful for the `Table` to visually represent the state that it is currently waiting on data to be loaded from the client. It can accomplish this fairly easily using existing state with no additional APIs or new components.
+
+Prototype visual:
+
+![Lazy Loading Spinner](./spec-images/LazyLoadingSpinner.gif)
+
+This ultimately will put the burden on the client to ensure that the `Table` is updated as needed to get rid of any displayed progress indicator, including in scenarios where the expansion of a parent row failed to load any children (possibly due to some client-side error). The `Table` will only guarantee that the progress indicator is shown when a parent row is expanded and it currently has no children, and that it will be removed once children are present.
+
+Expected user workflow:
+
+1. User loads data into table that specifies some rows as parents, but has no rows that indicate that row as a parent.
+2. User clicks on a parent row
+3. Row expands showing "row loading" indicator
+4. User handles event that row was expanded
+5. User sets data on table that has children for the row that was expanded
+6. "Row loading" indicator is removed and child rows are now displayed
+
+_Notes: When the new children are finally displayed, scroll position will be maintained in that the scroll index isn't adjusted. So a row that is currently displayed at the top of the table can be pushed down if new rows are added above it. Additionally, any previously expanded rows should remain expanded._
+
+Error workflow:
+
+1. Same as steps 1-4 above
+2. Error occurs retrieving data for child rows
+3. User calls `collapseRows(...)` on table instance passing parent row id that was previously expanded.
+4. Row is collapsed and "row loading" indicator is no longer displayed.
+5. Client should also provide appropriate error messaging within the application
+
+### Sorting
+
+Tanstack will sort children within each parent row by the same column that the parent is being sorted by. So, it will not be possible to sort children by a different column value than the one its parent is sorted by.
```

Comment:
To be very explicit about it: sorting has lower precedence than hierarchy. So sorting will only affect the order of rows at each level of hierarchy and only within each parent.

A consequence of this is that sorting won't necessarily show you the largest value at the top/bottom of the table.

I think all of that is sane, but it might be good to mention it explicitly.